doc/zmq_setsockopt.txt

```diff
@@ -1289,6 +1289,24 @@ Option value unit:: 0, 1
 Default value:: 1
 Applicable socket types:: ZMQ_RADIO, when using UDP multicast transport
 
+
+ZMQ_ROUTER_NOTIFY: Send connect and disconnect notifications
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Enable connect and disconnect notifications on a ROUTER socket.
+When enabled, the socket delivers a zero-length message (with routing-id
+as first frame) when a peer connects or disconnects. It's possible
```

Comment:
Is it ensured that connect/disconnect messages are always correctly paired?
Isn't it possible that either might be lost? Then it wouldn't be possible to correctly tell whether a notification is a connect or disconnect notification. In that case, it would be necessary not to deliver a zero-length message, but a longer message that allows to differentiate this.

Reply:
Well, I'm afraid my knowledge of libzmq internals doesn't let me confidently answer that. :( I tried to understand the stream_engine and follow the already existing STREAM_NOTIFY implementation that seemed to be stable. 

The connect notification is emitted right after the routing id is sent from the stream_engine to the router_socket, the disconnect is emitted on error only if the handshake is complete. These notifications are close to the corresponding monitor events, which (as I understand) are correctly paired. Apart from this, I can't provide more solid guarantees. :(

Regarding your proposal, yes, the notification messages could be longer the disambiguate between the two. This way the application could detect a malfunction in the library, but it won't prevent emitting incorrectly paired notifications, only improve detectability. I can add one byte with the ZMQ_NOTIFY_* flag values to the notification message if that looks better.

My main problem with the current design is that a malicious peer can craft messages identical to the notifications, but I couldn't come up with anything better, and tried to stick to what already exists (STREAM_NOTIFY).

Reply:
I think I was mistaken. The connect/disconnect messages are only sent locally, not via the underlying socket, right? Then they cannot be lost, of course.

However, this has another consequence, and you note that the disconnect message is only sent in case of an error. What happens in case of a normal connection termination? Or does at connection always end with an stream_engine_t::error?

I originally thought (not based on the implementation, but on this documentation), that the peer sends these messages over the network, in particular when it closes its socket. This would be useful IMO, because otherwise a disconnecting client cannot be reliably detected in time.

The same seems to be true for ZMQ_STREAM_NOTIFY, which may be stable, but I don't think it is widely used.

Reply:
The notification messages are created in the stream_engine and sent to the corresponding socket internally, these messages never appear "on the wire", contrary to ROUTER_PROBE messages.

AFAICT, all connection terminations end up in zmq::stream_engine_t::error(), as this is the single point in code where the 'disconnected' monitor event is emitted.

In my use-case I would like to be notified of the termination of a routing-id when a peer abruptly goes away, so my application can release the corresponding state (on a connection oriented transport). Disconnect messages are already present in the application layer protocol, but these are not sent when the app crashes, or forcefully killed.

Reply:
Ok, thanks for elaborating on this. I think there should be some comments in the code that clarify this, but this is beyond the scope of this PR.